api/v1alpha1/awsloadbalancercontroller_types.go

```diff
@@ -20,22 +20,85 @@ import (
+	AWSAddonWAFv2  AWSAddon = "AWSAddonWAFv2"
+)
+
+// +kubebuilder:validation:Enum=Auto;Manual
+type SubnetTaggingPolicy string
+
+const (
+
+	// AutoSubnetTaggingPolicy enables automatic subnet tagging.
+	AutoSubnetTaggingPolicy SubnetTaggingPolicy = "Auto"
+
+	// ManualSubnetTaggingPolicy disables automatic subnet tagging.
+	ManualSubnetTaggingPolicy SubnetTaggingPolicy = "Manual"
+)
 
 // AWSLoadBalancerControllerSpec defines the desired state of AWSLoadBalancerController
 type AWSLoadBalancerControllerSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
 
-	// Foo is an example field of AWSLoadBalancerController. Edit awsloadbalancercontroller_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// SubnetTagging describes how resource tagging will be done by the operator.
+	//
+	// When in "Auto", the operator will detect the subnets where the load balancers
+	// will be provisioned and have the required resource tags on them. Whereas when
+	// set to manual, this responsibility lies on the user.
+	//
+	SubnetTagging SubnetTaggingPolicy `json:"subnetTagging,omitempty"`
+
+	// Default AWS Tags that will be applied to all AWS resources managed by this
+	// controller (default []).
+	//
+	// This value is required so that this controller can function as expected
+	// in parallel to openshift-router.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	AdditionalResourceTags map[string]string `json:"additionalResourceTags,omitempty"`
+
+	// IngressClass specifies the Ingress class which the controller will reconcile.
+	// This will default to "alb".
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	IngressClass string `json:"ingressClass,omitempty"`
+
+	// Config specifies further customization options for the controller's deployment spec.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Config AWSLoadBalancerDeploymentConfig `json:"config,omitempty"`
+
+	// AWSAddon describes the AWS services that can be integrated with
+	// the AWS Load Balancer.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	EnabledAddons []AWSAddon `json:"enabledAddons,omitempty"` // indicates which AWS addons should be disabled.
+
 }
 
-// AWSLoadBalancerControllerStatus defines the observed state of AWSLoadBalancerController
+type AWSLoadBalancerDeploymentConfig struct {
+
+	// +kubebuilder:validation:Optional
+	// +optional
+	Replicas int `json:"replicas,omitempty"`
+}
+
+// AWSLoadBalancerControllerStatus defines the observed state of AWSLoadBalancerController.
 type AWSLoadBalancerControllerStatus struct {
-	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
+	// Conditions is a list of operator-specific conditions
+	// and their status.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
```

Comment:
Just to be consistent, let's add optional comment markers for `Conditions` and `ObservedGeneration`.

Reply:
fixed

Reply:
Cosmetic comment.
Sometimes we use `+optional` marker, sometimes both `+kubebuilder:validation:Optional` and `+optional`.
Still not quite sure whether we need to have both markers but just to keep things consistent: would you mind to align all the markers, @thejasn?

Reply:
Asked the question in kubebuilder channel of Kubernetes Slack. Both markers have the same effect in kubebuilder. `+kubebuilder:validation` seems to be kubebuilder specific while `+optional` is defined in [Kube API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md). So if the API is expected to be used by other code generators ([example](https://github.com/kubernetes/code-generator)): better to keep both markers. If only kubebuilder is expected to be used: `+kubebuilder:validation` is enough.

Reply:
I just realized after approving this PR that the markers are not really needed because the struct is not marked as required. So I've removed them [here](https://github.com/openshift/aws-load-balancer-operator/pull/9).